<a href="https://colab.research.google.com/github/sumukshashidhar/shamirs-secret-sharing-implementation/blob/master/SSS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Shamir's Secret Sharing

Shamir's Secret Sharing is an algorithm in cryptography created by Adi Shamir. It is a form of secret sharing, where a secret is divided into parts, giving each participant its own unique part.

To reconstruct the original secret, a minimum number of parts is required. In the threshold scheme this number is less than the total number of parts. Otherwise all participants are needed to reconstruct the original secret.

This is a basic implementation of the idea using integers instead of finite field arithmetic.

[Wiki Link](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)



# Basis

We have a secret

` secret = 90231` 

We want to divide this secret into a total of 3 parts, where any two parts are enough to reconstruct this integer

The total number of parts we wish to deconstruct the key into is $n$ and the number of keys that we require to reconstruct is $k$

In [6]:
secret = 90231
n = 3
k = 2

As suggested, we need to obtain $k-1$ random numbers to construct our polynomial

In [5]:
import random
r1 = random.randint(1, 10000)
r2 = random.randint(1, 10000)
print(r1, r2)

4042 7821


We have obtained the random numbers $r1=4042$ and $r2=7821$

Our **polynomial** therefore becomes $f(x) = 90231 + 4042x + 7821x^2$

We will now construct a total of three points from this polynomial

In [7]:
def get_points(x):
    return (secret + r1*x + r2*(x**2))


p1 = get_points(1)
p2 = get_points(2)
p3 = get_points(3)
print(p1, p2, p3)

102094 129599 172746


Great! We now have three points